test/test_common/simulated_time_system.cc

```diff
@@ -215,58 +215,77 @@ SimulatedTimeSystemHelper::~SimulatedTimeSystemHelper() { --instance_count; }
 bool SimulatedTimeSystemHelper::hasInstance() { return instance_count > 0; }
 
 SystemTime SimulatedTimeSystemHelper::systemTime() {
-  Thread::LockGuard lock(mutex_);
+  absl::MutexLock lock(&mutex_);
   return system_time_;
 }
 
 MonotonicTime SimulatedTimeSystemHelper::monotonicTime() {
-  Thread::LockGuard lock(mutex_);
+  absl::MutexLock lock(&mutex_);
   return monotonic_time_;
 }
 
-void SimulatedTimeSystemHelper::sleep(const Duration& duration) {
+void SimulatedTimeSystemHelper::advanceTimeAsync(const Duration& duration) {
   only_one_thread_.checkOneThread();
-  mutex_.lock();
+  absl::MutexLock lock(&mutex_);
   MonotonicTime monotonic_time =
       monotonic_time_ + std::chrono::duration_cast<MonotonicTime::duration>(duration);
-  setMonotonicTimeAndUnlock(monotonic_time);
+  setMonotonicTimeLockHeld(monotonic_time);
+}
+
+void SimulatedTimeSystemHelper::advanceTimeWait(const Duration& duration) {
+  only_one_thread_.checkOneThread();
+  absl::MutexLock lock(&mutex_);
+  MonotonicTime monotonic_time =
+      monotonic_time_ + std::chrono::duration_cast<MonotonicTime::duration>(duration);
+  setMonotonicTimeLockHeld(monotonic_time);
+  waitForNoPendingLockHeld();
+}
+
+void SimulatedTimeSystemHelper::waitForNoPendingLockHeld() const EXCLUSIVE_LOCKS_REQUIRED(mutex_) {
+  mutex_.Await(absl::Condition(
+      +[](const uint32_t* pending_alarms) -> bool { return *pending_alarms == 0; },
+      &pending_alarms_));
 }
 
 Thread::CondVar::WaitStatus SimulatedTimeSystemHelper::waitFor(
     Thread::MutexBasicLockable& mutex, Thread::CondVar& condvar,
     const Duration& duration) noexcept EXCLUSIVE_LOCKS_REQUIRED(mutex) {
   only_one_thread_.checkOneThread();
+
+  // TODO(#10568): This real-time polling delay should not be necessary. Without
+  // it, test/extensions/filters/http/cache:cache_filter_integration_test fails
+  // about 40% of the time.
   const Duration real_time_poll_delay(
       std::min(std::chrono::duration_cast<Duration>(std::chrono::milliseconds(50)), duration));
   const MonotonicTime end_time = monotonicTime() + duration;
 
-  while (true) {
+  bool cont = true;
```

Comment:
better naming please?

Reply:
Renamed to `timeout_not_reached`

Reply:
did you forget to upload?  I'm not seeing it?

Reply:
Nope; I thought I'd try intentionally not uploading until you were fine with my proposed changes, to avoid wasted CI churn :)